lint: unnamedResult should not warn on different named types

Reproducer:
```
gocritic check-package -enable=unnamedResult io
```
Output:
```
io/pipe.go:186:1: unnamedResult: consider to give name to results
```
Code:
```go
func Pipe() (*PipeReader, *PipeWriter) {
```
In my opinion, this signature is perfectly fine.
If there are distinct named types, type names serve as a "concise names".

Negative test suite should be extended as well.

--- a/lint/testdata/unnamedResult/negative_tests.go
+++ b/lint/testdata/unnamedResult/negative_tests.go
@@ -47,3 +47,52 @@ func g11() (x, y, z int) {
 func g12() (x, y int, _, w float64) {
 	return 0, 0, 0, 0
 }
+
+// This does violate Go conventions, but this is golint deal now.
+func errorBool() (error, bool) {
+	return nil, false
+}
+
+func boolError() (bool, error) {
+	return false, nil
+}
+
+func differentNamed() (namedInt, namedStruct) {
+	return 0, namedStruct{}
+}
+
+func differentNamedAndError() (namedInt, namedStruct, error) {
+	return 0, namedStruct{}, nil
+}
+
+func differentNamedAndBool() (namedInt, namedStruct, bool) {
+	return 0, namedStruct{}, false
+}
+
+func namedSlices() ([]namedInt, []namedStruct) {
+	return nil, nil
+}
+
+func namedSlices2() ([][]namedInt, []namedStruct) {
+	return nil, nil
+}
+
+func namedSlices3() ([][]namedInt, [][][]namedStruct) {
+	return nil, nil
+}
+
+func namedArrays() ([2]namedInt, [4]namedStruct) {
+	return [2]namedInt{}, [4]namedStruct{}
+}
+
+func namedPointers() (*namedInt, *namedStruct) {
+	return nil, nil
+}
+
+func namedPointers2() (**namedInt, *namedStruct) {
+	return nil, nil
+}
+
+func namedPointers3() (**namedInt, ***namedStruct) {
+	return nil, nil
+}

--- a/lint/testdata/unnamedResult/positive_tests.go
+++ b/lint/testdata/unnamedResult/positive_tests.go
@@ -51,3 +51,32 @@ func (f *foo) f9() (bool, func() int) {
 func f10() (int, int, float64, float64) {
 	return 0, 0, 0, 0
 }
+
+/// consider to give name to results
+func doubleError() (error, error) {
+	return nil, nil
+}
+
+type namedInt int
+
+type namedStruct struct{}
+
+/// consider to give name to results
+func named2ptr() (namedInt, *namedInt) {
+	return 0, nil
+}
+
+/// consider to give name to results
+func named2() (namedInt, namedInt) {
+	return 0, 0
+}
+
+/// consider to give name to results
+func named3() (namedInt, namedStruct, namedInt) {
+	return 0, namedStruct{}, 0
+}
+
+/// consider to give name to results
+func namedAndPrimitive() (namedInt, int) {
+	return 0, 0
+}

--- a/lint/unnamedResult_checker.go
+++ b/lint/unnamedResult_checker.go
@@ -25,48 +25,62 @@ type unnamedResultChecker struct {
 func (c *unnamedResultChecker) VisitFuncDecl(decl *ast.FuncDecl) {
 	results := decl.Type.Results
 	switch {
-	case results == nil || results.NumFields() < 2:
-		return
+	case results == nil:
+		return // Function has no results
+	case len(results.List) > 0 && results.List[0].Names != nil:
+		return // Skip named results
+	}
 
-	case results.NumFields() == 2:
-		typ1, typ2 := c.getResultsTypes(results.List)
-		if len(results.List[0].Names) == 2 ||
-			(!c.isBoolOrError(typ1) && c.isBoolOrError(typ2)) {
-			// no need to name results for (T, error) or (T, bool)
-		} else {
+	typeName := func(x ast.Expr) string { return c.typeName(c.ctx.typesInfo.TypeOf(x)) }
+	isError := func(x ast.Expr) bool { return qualifiedName(x) == "error" }
+	isBool := func(x ast.Expr) bool { return qualifiedName(x) == "bool" }
+
+	// Main difference with case of len=2 is that we permit any
+	// typ1 as long as second type is either error or bool.
+	if results.NumFields() == 2 {
+		typ1, typ2 := results.List[0].Type, results.List[1].Type
+		name1, name2 := typeName(typ1), typeName(typ2)
+		cond := (name1 != name2 && name2 != "") ||
+			(!isError(typ1) && isError(typ2)) ||
+			(!isBool(typ1) && isBool(typ2))
+		if !cond {
 			c.warn(decl)
 		}
+		return
+	}
 
-	default:
-		for _, res := range results.List {
-			if len(res.Names) == 0 {
-				c.warn(decl)
-				break
-			}
+	seen := make(map[string]bool, len(results.List))
+	for i := range results.List {
+		typ := results.List[i].Type
+		name := typeName(typ)
+		isLast := i == len(results.List)-1
+
+		cond := !seen[name] ||
+			(isLast && (isError(typ) || isBool(typ)))
+		if !cond {
+			c.warn(decl)
+			return
 		}
+
+		seen[name] = true
+	}
+}
+
+func (c *unnamedResultChecker) typeName(typ types.Type) string {
+	switch typ := typ.(type) {
+	case *types.Array:
+		return c.typeName(typ.Elem())
+	case *types.Pointer:
+		return c.typeName(typ.Elem())
+	case *types.Slice:
+		return c.typeName(typ.Elem())
+	case *types.Named:
+		return typ.Obj().Name()
+	default:
+		return ""
 	}
 }
 
 func (c *unnamedResultChecker) warn(n ast.Node) {
 	c.ctx.Warn(n, "consider to give name to results")
-}
-
-func (c *unnamedResultChecker) getResultsTypes(fields []*ast.Field) (res1, res2 ast.Expr) {
-	if len(fields) == 2 {
-		return fields[0].Type, fields[1].Type
-	}
-	return fields[0].Type, fields[0].Type
-}
-
-func (c *unnamedResultChecker) isBoolOrError(expr ast.Expr) bool {
-	switch typ := c.ctx.typesInfo.TypeOf(expr).(type) {
-	case *types.Named:
-		return typ.Obj().Name() == "error"
-
-	case *types.Basic:
-		return typ.Kind() == types.Bool
-
-	default:
-		return false
-	}
 }

